src/util/codec/mysql/json/json_modify.rs

```diff
@@ -0,0 +1,175 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::Json;
+use super::super::Result;
+use super::path_expr::{PathLeg, PathExpression, contains_any_asterisk};
+
+/// `ModifyType` is for modify a JSON.
+#[derive(Clone, Debug)]
+pub enum ModifyType {
+    /// `Insert` is for inserting a new element into a JSON.
+    Insert,
+    /// `Replace` is for replacing a old element from a JSON.
+    Replace,
+    /// `Set` = `Insert` | `Replace`
+    Set,
+}
+
+impl Json {
+    // Modifies a Json object by insert, replace or set.
+    // All path expressions cannot contain * or ** wildcard.
+    // If any error occurs, the input won't be changed.
+    pub fn modify(self,
+                  path_expr_list: &[PathExpression],
+                  mut values: Vec<Json>,
+                  mt: ModifyType)
+                  -> (Json, Result<()>) {
+        if path_expr_list.len() != values.len() {
+            return (self, Err(box_err!("Incorrect parameter count")));
+        }
+        for expr in path_expr_list {
+            if contains_any_asterisk(expr.flags) {
+                return (self, Err(box_err!("Invalid path expression")));
```

Comment:
It seems mysql-5.7.18 returns some other details in the similar case as following
```
mysql> SELECT JSON_SET(@j, '$[1].b[0]', 1, '$[1].*', 2);
ERROR 3149 (42000): In this situation, path expressions may not contain the * and ** tokens.
```